File: internal/data/appdata.go
package data

import (
	"encoding/json"
	"fmt"
	"os"
	"sort"
	"strings"

	"github.com/ball6847/modelsdb/internal/models"
)

type AppData struct {
	Models []models.Model
}

func Load(path string) (*AppData, error) {
	data, err := os.ReadFile(path)
	if err != nil {
		return nil, fmt.Errorf("failed to read %s: %w", path, err)
	}

	var raw map[string]json.RawMessage
	if err := json.Unmarshal(data, &raw); err != nil {
		return nil, fmt.Errorf("failed to parse api.json: %w", err)
	}

	var allModels []models.Model
	for providerID, providerData := range raw {
		var provider struct {
			Models map[string]models.Model `json:"models"`
		}
		if err := json.Unmarshal(providerData, &provider); err != nil {
			continue
		}
		for _, m := range provider.Models {
			m.ProviderID = providerID
			allModels = append(allModels, m)
		}
	}

	return &AppData{Models: allModels}, nil
}

func (a *AppData) Filter(query string) []models.Model {
	if query == "" {
		return a.Models
	}
	q := strings.ToLower(query)
	var result []models.Model
	for _, m := range a.Models {
		if strings.Contains(strings.ToLower(m.Name), q) ||
			strings.Contains(strings.ToLower(m.ID), q) ||
			strings.Contains(strings.ToLower(m.ProviderID), q) ||
			(m.Family != nil && strings.Contains(strings.ToLower(*m.Family), q)) {
			result = append(result, m)
		}
	}
	return result
}

func (a *AppData) Sort(items []models.Model, sortBy models.SortField, dir models.SortDir) []models.Model {
	sorted := make([]models.Model, len(items))
	copy(sorted, items)

	sort.SliceStable(sorted, func(i, j int) bool {
		var vi, vj float64
		var ni, nj bool

		switch sortBy {
		case models.SortName:
			return sortDirLess(sorted[i].Name < sorted[j].Name, dir)
		case models.SortProvider:
			return sortDirLess(sorted[i].ProviderID < sorted[j].ProviderID, dir)
		case models.SortContext:
			ni, nj = sorted[i].Limit == nil, sorted[j].Limit == nil
			if !ni {
				vi = float64(sorted[i].Limit.Context)
			}
			if !nj {
				vj = float64(sorted[j].Limit.Context)
			}
		case models.SortInputCost:
			ni, nj = sorted[i].Cost == nil, sorted[j].Cost == nil
			if !ni {
				vi = sorted[i].Cost.Input
			}
			if !nj {
				vj = sorted[j].Cost.Input
			}
		case models.SortOutputCost:
			ni, nj = sorted[i].Cost == nil, sorted[j].Cost == nil
			if !ni {
				vi = sorted[i].Cost.Output
			}
			if !nj {
				vj = sorted[j].Cost.Output
			}
		default:
			return sortDirLess(sorted[i].Name < sorted[j].Name, dir)
		}

		return sortDirLessWithNil(vi, vj, ni, nj, dir)
	})

	return sorted
}

func sortDirLess(less bool, dir models.SortDir) bool {
	if dir == models.SortDesc {
		return !less
	}
	return less
}

func sortDirLessWithNil(vi, vj float64, ni, nj bool, dir models.SortDir) bool {
	if ni && nj {
		return false
	}
	if ni {
		return false // i is nil → goes to end
	}
	if nj {
		return true // j is nil → i comes first
	}
	if dir == models.SortDesc {
		return vi > vj
	}
	return vi < vj
}

func (a *AppData) Find(providerID, modelID string) (*models.Model, error) {
	for _, m := range a.Models {
		if m.ProviderID == providerID && m.ID == modelID {
			return &m, nil
		}
	}
	return nil, fmt.Errorf("model not found: %s/%s", providerID, modelID)
}

func Paginate(items []models.Model, offset, limit int) []models.Model {
	if offset >= len(items) {
		return []models.Model{}
	}
	end := offset + limit
	if end > len(items) {
		end = len(items)
	}
	return items[offset:end]
}

File: server/services/app-data.ts
import type { Model, ModelPage, SortField, SortDir } from "../types/model.ts";

export class AppData {
  models: Model[];

  constructor(models: Model[]) {
    this.models = models;
  }

  replace(models: Model[]) {
    this.models = models;
  }
}

export async function load(path: string): Promise<AppData> {
  const data = await Deno.readTextFile(path);
  const raw: Record<string, { models: Record<string, Model> }> = JSON.parse(data);

  const allModels: Model[] = [];
  for (const [providerID, providerData] of Object.entries(raw)) {
    if (!providerData.models) continue;
    for (const m of Object.values(providerData.models)) {
      allModels.push({ ...m, provider_id: providerID });
    }
  }

  return new AppData(allModels);
}

export function filter(data: AppData, query: string): Model[] {
  if (!query) return data.models;
  const q = query.toLowerCase();
  return data.models.filter((m) =>
    m.name.toLowerCase().includes(q) ||
    m.id.toLowerCase().includes(q) ||
    m.provider_id.toLowerCase().includes(q) ||
    (m.family && m.family.toLowerCase().includes(q))
  );
}

export function sort(items: Model[], sortBy: SortField, dir: SortDir): Model[] {
  const sorted = [...items];

  sorted.sort((a, b) => {
    switch (sortBy) {
      case "name":
        return sortDirLess(a.name < b.name, a.name === b.name, dir);
      case "provider":
        return sortDirLess(a.provider_id < b.provider_id, a.provider_id === b.provider_id, dir);
      case "context":
        return sortDirLessWithNil(
          a.limit?.context,
          b.limit?.context,
          dir,
        );
      case "input_cost":
        return sortDirLessWithNil(
          a.cost?.input,
          b.cost?.input,
          dir,
        );
      case "output_cost":
        return sortDirLessWithNil(
          a.cost?.output,
          b.cost?.output,
          dir,
        );
      default:
        return sortDirLess(a.name < b.name, a.name === b.name, dir);
    }
  });

  return sorted;
}

function sortDirLess(less: boolean, equal: boolean, dir: SortDir): number {
  if (equal) return 0;
  if (dir === "desc") return less ? 1 : -1;
  return less ? -1 : 1;
}

function sortDirLessWithNil(vi: number | undefined, vj: number | undefined, dir: SortDir): number {
  const ni = vi === undefined;
  const nj = vj === undefined;
  if (ni && nj) return 0;
  if (ni) return 1;
  if (nj) return -1;
  if (dir === "desc") return vj! - vi!;
  return vi! - vj!;
}

export function find(data: AppData, providerID: string, modelID: string): Model | undefined {
  return data.models.find((m) => m.provider_id === providerID && m.id === modelID);
}

export function paginate(items: Model[], offset: number, limit: number): Model[] {
  if (offset >= items.length) return [];
  const end = Math.min(offset + limit, items.length);
  return items.slice(offset, end);
}

export function buildPage(items: Model[], offset: number, limit: number): ModelPage {
  const total = items.length;
  const page = paginate(items, offset, limit);
  return { models: page, total };
}
